controllers/template_sync.go

```diff
@@ -296,32 +283,48 @@ func (r *PolicyReconciler) Reconcile(ctx context.Context, request reconcile.Requ
 		if (!equality.Semantic.DeepEqual(eObjectUnstructured["spec"], tObjectUnstructured.Object["spec"])) ||
 			(!equality.Semantic.DeepEqual(eObject.GetAnnotations(), tObjectUnstructured.GetAnnotations())) {
 			// doesn't match
-			reqLogger.Info("Existing object and template didn't match, will update", "PolicyTemplateName", tName)
+			tLogger.Info("Existing object and template didn't match, will update")
 
 			eObjectUnstructured["spec"] = tObjectUnstructured.Object["spec"]
 
 			eObject.SetAnnotations(tObjectUnstructured.GetAnnotations())
 
 			_, err = res.Update(ctx, eObject, metav1.UpdateOptions{})
 			if err != nil {
-				reqLogger.Error(err, "Failed to update the policy template", "PolicyTemplateName", tName)
-				r.Recorder.Event(instance, "Warning", "PolicyTemplateSync",
-					fmt.Sprintf("Failed to update policy template %s", tName))
+				resultError = err
+				errMsg := fmt.Sprintf("Failed to update policy template %s: %s", tName, err)
+
+				r.emitTemplateError(instance, tIndex, tName, errMsg)
+				tLogger.Error(err, "Failed to update the policy template")
+
+				continue
+			}
 
-				return reconcile.Result{}, err
+			successMsg := fmt.Sprintf("Policy template %s was updated successfully", tName)
+
+			err = r.handleSyncSuccess(ctx, instance, tIndex, tName, successMsg, res)
+			if err != nil {
+				resultError = err
+				tLogger.Error(resultError, "Error after updating template (will requeue)")
 			}
 
-			reqLogger.Info("Existing object has been updated", "PolicyTemplateName", tName)
-			r.Recorder.Event(instance, "Normal", "PolicyTemplateSync",
-				fmt.Sprintf("Policy template %s was updated successfully", tName))
+			tLogger.Info("Existing object has been updated")
 		} else {
-			reqLogger.Info("Existing object matches the policy template", "PolicyTemplateName", tName)
+			successMsg := fmt.Sprintf("Policy template %s matches the existing object", tName)
+
+			err = r.handleSyncSuccess(ctx, instance, tIndex, tName, successMsg, res)
```

Comment:
Good point with the status update causing a reconcile, a predicate will be good.

You're correct, `handleSyncSuccess` will keep wiping the template object's complianceState (possibly a no-op) until the status is updated via a new compliance event from the relevant policy-controller. I would rather have cases where that happens multiple times than cases where that is missed.

I wasn't thinking about the `"Normal", "PolicyTemplateSync"` events being too much, but they definitely will be spammed in policies with multiple templates, which is not good. What if we just skip those events in this "happy-path" where the template already matches? It would just check for `template-error` and handle that like it currently does.

Reply:
Your suggestion sounds good.